Conjugate view tensor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44799 Conjugate view tensor**

Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Conjugate view tensor

Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
cc @boeddeker

albanD commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,97 @@
+  //     call it a day.
+  //  2. Inplace operation.  Desugar x.add_(2) into x.conj_().add_(2).conj_().
+  //     Materialize other inputs as in (1).
+  //  3. Out-of-place operation.  Desugar add(x, 2, out=y) into y.copy_(add(x, 2))
```
nit: This is not really Out of place. Inplace via out? haha

albanD commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -200,17 +201,30 @@ Tensor imag(const Tensor& self) {
-  }
-  return at::_conj(self);
+  Tensor self_;
+  auto impl = c10::make_intrusive<TensorImpl>(
```
Is this expected to be different from `self.variable_data()` + setting conj?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1103,11 +1103,16 @@
 - name: view(Tensor(a) self, int[] size) -> Tensor(a)
   self: grad.reshape(self.sizes())
 
+# When result1 is true, this is effectively a fused conj() and view_as_real
```
There is no result1 here?

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -315,6 +315,11 @@ def saved_variables(formula, args):
             'type': 'int64_t',
         }),
+        # replace self.is_conj() with self_conjugate
+        (r'{}.is_conj\(\)', {
```
Is this always a valid thing to do? Is there any case where this flag can be changed inplace on a Tensor?

ezyang replied:
Yes it's always valid. Even if the flag is changed inplace, the old saved value is the correct one to be used for backwards, similar to how you can resize a tensor that was saved for backwards, but you'll get the old size in the backwards itself (and `resize_` doesn't actually do a VC update)

boeddeker commented:
Here, just some of my thoughts:
 - Is it possible for a user to easily see what happens? It sounds like many magic in the background (e.g. when does it work, that conj was not explicit executed or how does a user know if a particular operation handles the conj flag or uses the fallback).
 - What happens for the backward pass? Is the non conjugate and the conjugate kept in the memory (e.g. `x.conj() * x`)?
 - How about adding a negative flag? It is similar to the conjugate and it would simplify the accessing of the imag property. I am not sure how often this happens, but it could have also a small positive effect for real operations.
 - `x.view_as_real_physical()`: I am not sure, what this function does, but is with `real` the real part of a complex number meant (I expect not)? If not, I would suggest finding a name, that does not contain the word `real` (It may lead to confusion). When working with complex numbers, IMHO the words complex and real should be reserved for the complex numbers. 
 - `x.resolve_conj().imag`: This is a confusing call. It would lead to much boilerplate code and unexpected behavior. I know "Explicit is better than implicit", but this is an implicit technique and making some parts implicit and some explicit is confusing.
 - What happens, when an operation uses the fallback and a tensor is used two times (e.g. `y = x.conj(); z = y * y` (Assume `__mul__` would not support the conj flag)). Is then the conjugate calculated two times?

ezyang commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,97 @@
+
+  for (int64_t i = 0; i < num_arguments; ++i) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (!ivalue.isTensor()) {
```
Not handled, this is a bug.

ezyang commented:
> Is it possible for a user to easily see what happens? It sounds like many magic in the background (e.g. when does it work, that conj was not explicit executed or how does a user know if a particular operation handles the conj flag or uses the fallback).

I don't know a good way to surface this. The semantics of `conj` are fairly simple: it *never* does work, it always operates lazily. But whether or not you managed to hit a fused kernel or not is not something that I really know how to surface naturally. I think in the past we've talked about having some sort of logging to let people know when you hit a slow path, and you can also find things about this by running the autograd profiler (or at least, you *should* be able to); however, giving a warning every time we explicitly materialize a conj will probably be too much for most users.

> What happens for the backward pass? Is the non conjugate and the conjugate kept in the memory (e.g. x.conj() * x)?

Only one is kept! A big benefit, thanks for telling us about it :)

> How about adding a negative flag? It is similar to the conjugate and it would simplify the accessing of the imag property. I am not sure how often this happens, but it could have also a small positive effect for real operations.

Yes, between this and your comment about `x.resolve_conj().imag`, it sounds like we should directly implement negative flag today, rather than waiting later. I guess it doesn't have to be efficient or particularly well integrated, which means that the code here wouldn't be too large. If we did this, then your `x.resolve_conj()` problem would also be solved.

> x.view_as_real_physical(): I am not sure, what this function does, but is with real the real part of a complex number meant (I expect not)? If not, I would suggest finding a name, that does not contain the word real (It may lead to confusion). When working with complex numbers, IMHO the words complex and real should be reserved for the complex numbers.

Ah, this is a more general API problem. `view_as_real()` is our current naming for reinterpreting a C tensor as a R^2 tensor; so `view_as_real_physical()` does the same thing but disregarding the conjugation status (because I don't have a way to do a view in R^2 but only have every other column be negative.) Do you think `view_as_r2` would be clearer?

> What happens, when an operation uses the fallback and a tensor is used two times (e.g. y = x.conj(); z = y * y (Assume __mul__ would not support the conj flag)). Is then the conjugate calculated two times?

Yes, conjugate is calculated two times, unless the operation supported fusion.

boeddeker commented:
Thank you for the answers.

> I don't know a good way to surface this. The semantics of conj are fairly simple: it never does work, it always operates lazily. But whether or not you managed to hit a fused kernel or not is not something that I really know how to surface naturally. I think in the past we've talked about having some sort of logging to let people know when you hit a slow path, and you can also find things about this by running the autograd profiler (or at least, you should be able to); however, giving a warning every time we explicitly materialize a conj will probably be too much for most users.

Thanks for the explanation. Yes, letting the user know what happened is difficult. Warnings are maybe too verbose and difficult to track.
How about exposing this information with the help of the backward graph?
When you visualize the backward graph and it shows the fused operations, you get an idea how good your code is.
(Afaik has pytorch at the moment no official tool to visualize the backward graph: https://discuss.pytorch.org/t/print-autograd-graph/692/7).

> Only one is kept! A big benefit, thanks for telling us about it :)

That is great.

> Yes, between this and your comment about x.resolve_conj().imag, it sounds like we should directly implement negative flag today, rather than waiting later. I guess it doesn't have to be efficient or particularly well integrated, which means that the code here wouldn't be too large. If we did this, then your x.resolve_conj() problem would also be solved.

I am not sure, where it would be used. Depending on this, it may have to be efficient.
When it is only set with `.conj().imag` or something like `set_neg` it has no effect to existing code, but changing `- tensor` may change existing code.
Nevertheless, adding this would lead to a cleaner complex API.

> Ah, this is a more general API problem. view_as_real() is our current naming for reinterpreting a C tensor as a R^2 tensor; so view_as_real_physical() does the same thing but disregarding the conjugation status (because I don't have a way to do a view in R^2 but only have every other column be negative.) Do you think view_as_r2 would be clearer?

Ah, now I get the intention for the name. It is probably difficult to find the right name for this.
In the past I used `view_as_float`, since numpy is limited to float types, when using complex numbers.
Something like `view_complex_as_r2_vector` or `view_complex_as_real_vector` would be more verbose, but I am not sure, if they are better.

> Yes, conjugate is calculated two times, unless the operation supported fusion.

So this is a small drawback of the current implementation.
I was hoping that the "python" variable holds the non-conjugate tensor and a lazy attribute that is used when an operation doesn't support fusion. But I am not familiar with the implementation of the python tensor. So it is probably impossible with the design.

ezyang commented:
> I was hoping that the "python" variable holds the non-conjugate tensor and a lazy attribute that is used when an operation doesn't support fusion. But I am not familiar with the implementation of the python tensor. So it is probably impossible with the design.

Yeah, this design explicitly duplicates work. One way this is observable, is that if you do an inplace operation on a conjugate view, the underlying base gets updated too. Lazy semantics wouldn't give you that.

The lazy attribute is a valid alternative in the design space, but it is not too easy to implement correctly in a multithreaded setting. The problem is that `y = x.conj()` may be accessed concurrently from multiple threads, and we must ensure that the instantiation of the lazy tensor doesn't cause race condition between the threads, which implies some sort of synchronization. It is known how to do this efficiently (e.g. Section 8 of https://www.microsoft.com/en-us/research/wp-content/uploads/2009/09/multicore-ghc.pdf ) but conjugate views avoided this problem entirely, so I had guessed that maybe we should do it instead. Wondering if this triggered any further thoughts.

boeddeker commented:
> Yeah, this design explicitly duplicates work. One way this is observable, is that if you do an inplace operation on a conjugate view, the underlying base gets updated too. Lazy semantics wouldn't give you that.

I think my idea is slightly different from yours.
When I get your idea correct, the python object holds only one data object (I think on Pytorch it is called storage).
And this may be changed inplace, when one operation doesn't support fusion.

My idea was slightly different, the python object can hold two data objects:
```python
class Tensor:
    def __init__(self, data):
        self.data = data
        self._conj_data = None
    @property
    def conj_data(self):
        if self._conj_data is None:
            self._conj_data = self.data.conj()
        return self._conj_data
```
In python this is thread save, because the assignment is atomic (Thanks to the GIL) and the garbage collector takes care of a potential race condition (i.e. delete the object that was first assigned to `self._conj_data`).

> The lazy attribute is a valid alternative in the design space, but it is not too easy to implement correctly in a multithreaded setting. The problem is that y = x.conj() may be accessed concurrently from multiple threads, and we must ensure that the instantiation of the lazy tensor doesn't cause race condition between the threads, which implies some sort of synchronization. It is known how to do this efficiently (e.g. Section 8 of https://www.microsoft.com/en-us/research/wp-content/uploads/2009/09/multicore-ghc.pdf ) but conjugate views avoided this problem entirely, so I had guessed that maybe we should do it instead. Wondering if this triggered any further thoughts.

I was thinking about the "PostCheck" idea (Without knowing that name).
How likely is it, that you use a conjugate tensor multiple times and how likely, that you use it in different threads.
I think PyTorch does the following (Sorry, that I am not so familiar with the core of PyTorch):
Python triggers a sequential workload, that is sequentially executed (maybe async, but sequentially).
For the backward graph I am not sure, but when it is also sequentially, the potential duplicate execution of the conjugate operation when using "PostCheck" doesn't matter, because it rarely (or never) happens.

ezyang commented:
> My idea was slightly different, the python object can hold two data objects:

Yep. Though, aren't you worried about memory usage? This suggests that when `_conj_data` gets filled, we should delete `data`, so that we don't force both the old data and the new data to be in memory.

> How likely is it, that you use a conjugate tensor multiple times and how likely, that you use it in different threads.

I agree, the multithreaded situation is unlikely to happen. Most of my implementation difficulty is more to do with correctness: we must not segfault or data race when you use it from multiple threads. BTW I realized that https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.68.8868&rep=rep1&type=pdf is a better paper on the same subject as above (see Section 3).

The primary challenge seems to be that we have arbitrarily many codepaths that may observe the `_conj_data`. In particular, all operations that do fused conjugation operations will interact directly with `_conj_data`. So unlike in the paper I cited, where we only had to worry about data races between two threads that were going to evaluate the thunk and then write it back, we have to handle data races between all materialization operations, as well as operators which know how to fuse the conjugation. In particular, when is it safe to delete the old `_conj_data`? It's not safe to delete it at the end of a materialization operation, as there may be other fused kernels that are still accessing `_conj_data` during the process.

Here is something that might work:

* In GHC, after a thunk is updated it is replaced by an indirection. This is required in their setting because in general the result data may not fit in the memory allocated to the thunk. In PyTorch, things are different because we have a uniform data representation for Tensors, so we can plausibly store all the information we need directly in TensorImpl.  So instead, a conjugate tensor is a regular tensor that also records the original tensor (so we can do the operation). We should also store the version counter of the base tensor, so that we can detect when a mutation on the base tensor invalidates our conjugate reference.
* We have to handle the data race between materializing a conjugate tensor, and a separate use of the conjugate tensor directly. There are two pieces of data we have to synchronize: the Tensor key set (which controls which kernel we dispatch to) and the intrusive pointer to the original tensor. Here is one possible way to handle the race:
  * Our first "test" is the presence of a conjugate bit in the key set. If it is present, we get dispatched to a different kernel. The conjugate bit is a conservative tests: a tensor may have already been materialized even though its conjugate bit is set; however, a tensor is guaranteed to be materialized if the conjugate bit is not set.
  * When we get into a fused kernel / materialize conjugate kernel, we now must do a second test for the conjugation status. This is done by weak intrusive pointer `lock()` on the original tensor; if successful, we (1) confirm that the tensor is conjugate, and (2) increment the refcount and get an owning reference to the unconjugated storage. We can now proceed to do the operation on the unconjugated version of the tensor. A test for the version counter should also occur at this point.
  * For materialize conjugate kernel, we first (1) populate the storage field of the tensor with the real value (this is non-atomic), (2) free the owning reference to the base tensor, (3) update the key set to delete the conjugate bit.
* This scheme is not directly compositional with other lazy schemes. If we want it to be compositional, we would probably have to introduce a generic `materialize()` operation, and then force a materialization before we actually do our operation in question.

I must admit, while I believe this could work, and I believe that there may be other (non-conjugate) use cases that would benefit greatly from this model, I am unconvinced that this is the best choice for conjugate tensors. But we are kind of at a fork in the road regarding semantics: the semantics of this and conjugate views are *different.* In particular:

```
x = ... some complex tensor ...
y = x.conj()
y.copy_(z)
```

For conjugate lazy tensors, you would expect `x` to stay unchanged. However, for conjugate view tensors, `x` is changed.

Also, writing this example has made me realize that it is not sound to allow duplicate work if the user subsequently mutates the result of the conjugation, since two copies of the conjugate tensor is observable. So you really are going to have to proper locking when you materialize.

boeddeker commented:
> Yep. Though, aren't you worried about memory usage? This suggests that when _conj_data gets filled, we should delete data, so that we don't force both the old data and the new data to be in memory.

Yes, I care about the memory, but mainly about the memory of the tensors hold in the backward graph.
It is non-trivial to remove them from the backward graph. In the forward pass it is easy to remove unused tensors.

When `_conj_data` is deleted after the lifetime of the python object, I don't see a clear overhead of this idea.
Yes, it consumes more data, because the original data is not deleted, but when would it be deleted?

I scrolled to some of my code and most of the time, the original data will be hold by another tensor.
So it cannot be deleted, while a forward operation uses the conjugated tensor.

Another thing you may want to consider is, that the code should not have unpredictable side effects:
```python
y = x.conj()
if calculate_z:
     z = y * x   # Assume mul has no fusion of conj and mul
w = y @ x  # Assume matmul has fusion of conj and matmul implemented
```
Depending, whether `z` is calculated, the `matmul` operations is a fused operation or not.
And independent of the calculation of `z`, the memory consumption does not change.
The original data is kept.

For the following, I am not sure, what happens.
Assume again your proposal for the inplace change, and the above example, where z and w is calculated.
When the matmul operation doesn't use the fused operation, because the mul operation triggerd the materialization,
the gradient calculation is less optimal, because it introduces one more intermediate gradient.
Here I am not sure, if the gradient will be materialized or not (i.e. if it is an overhead or not).

> I agree, the multithreaded situation is unlikely to happen. Most of my implementation difficulty is more to do with correctness: we must not segfault or data race when you use it from multiple threads.

I agree, avoiding segfault or data race should have the highest priority.
But making it efficient is not so important as long no error happens.

> The primary challenge seems to be that we have arbitrarily many codepaths that may observe the _conj_data. In particular, all operations that do fused conjugation operations will interact directly with _conj_data.

The idea was the opposite. In the long term (maybe very long) no operation will see _conj_data and they use the non-conjugate data and apply the fused operation. The fused operation should always be faster (e.g. for cuda: one less kernel call).

When I understand you right, your biggest problem is, that you have two possible realizations for the data:
`x` and `x.conj()`. And you try to predict, how to exchange them, without informing the user.
I think this is a very difficult task and maybe cannot be solved.

I would first do some investigations, if there is an improvement, to only keep 1 realization of the data.
I expect holding both gives more advantages, because you most likely hold the original data also somewhere else.
When the user cares about the higher memory usage, a non-lazy conjugate can do the task.

Furthermore, (at least at python level) just holding different realizations is simple and only need minimal locking (In Python the GIL can do this).

At the moment, I am unsure, if this implicit implementation of conj is better than exposing a conjugate argument.
It seems quite complicated and has many side effects.

ezyang commented:
> Yes, I care about the memory, but mainly about the memory of the tensors hold in the backward graph.
> It is non-trivial to remove them from the backward graph. In the forward pass it is easy to remove unused tensors.

This is good to know. To put it differently, if you have `x = y.conj()` and you care about not having either in memory, you will simply arrange for both x and y to get deleted. No problem. But then if x and y both get captured by the backwards pass (because they were used as part of some other computations), now you have no good solution, since there's no way to tell backward "hey, don't capture both of x and y".

> I scrolled to some of my code and most of the time, the original data will be hold by another tensor.

If it's public, linking it from this discussion would be very helpful :)

> Another thing you may want to consider is, that the code should not have unpredictable side effects:

> ...

> Here I am not sure, if the gradient will be materialized or not (i.e. if it is an overhead or not).

In the scheme I described in my post, it would be materialized aka have overhead. In contrast, conjugate views wouldn't have this problem.

I thought about your solution to repair this problem. I think this sentence sums it up nicely:

> When I understand you right, your biggest problem is, that you have two possible realizations for the data: x and x.conj(). And you try to predict, how to exchange them, without informing the user.
I think this is a very difficult task and maybe cannot be solved.

We can describe my two approaches in terms of how they make predictions about this exchange:

* **Conjugate view tensors** (as proposed by this PR) say, "Always keep x until the very latest moment, when you have to do the computation and there is no more time to wait." In PL terms, we'd say this is call-by-need evaluation.
* **Conjugate lazy tensors** (as described in my comment) say, "Keep x until the first time we need to materialize it; then swap it with x.conj() and use that from then on." This is lazy evaluation.

I think you are advancing the following third proposal:

* **Conjugate cached tensors**: "Always keep x. If we need to materialize it, compute x.conj, and keep that too. Use either x or x.conj depending on whether or not the operator supports fused operations, but preferring x." In the long term, no one needs to materialize (BTW, this is unrealistic, IMO), and so x.conj ends up never getting materialized.

In this third proposal, we also have to decide whether or not you have view semantics (mutating x.conj() affects x) or clone semantics (mutating x.conj() has no effect on x). Due to the materialized x.conj(), view semantics would be difficult, although still possible (using version counters). But it is still very tricky. For example, in your example, if `z` participated in backwards, then you would have ended up with both `x` and `x.conj()` saved for backwards (indeed, only conjugate views save you, at the cost of a bit more recompute).

> Furthermore, (at least at python level) just holding different realizations is simple and only need minimal locking (In Python the GIL can do this).

Alas, I need this to work in C++ only XD. This is one place where you probably don't care, but for the good of the framework we need to make sure this is handled correctly.

> At the moment, I am unsure, if this implicit implementation of conj is better than exposing a conjugate argument.

Yes. This is something that I am worried about too. Right now, I think it is worth it, but mostly because I think the mathematicians in the audience will get really irritated at having to translate all of their code into using explicit conjugate flags. Chances are they'll just make some wrapper object to handle it... and if they're going to do that, we might as well do it properly for them in framework.

I do want to reiterate, that I feel that lazy conjugate views (which are implemented in this PR) are a lot simpler than lazy conjugate tensors. Their primary downside is duplicated work materializing conjugate, but as you say, if in the long term we add fusion to most operators, you will no longer have duplicated work. (Also, I suspect one could always add a cache on top later as a performance optimization; though we do have to make a call on clone or view semantics sooner rather than later.) What do you think?

boeddeker commented:
Thanks, your explanation for conjugate view tensors and conjugate lazy tensors helped me. 
I mixed them.

> This is good to know. To put it differently, if you have x = y.conj() and you care about not having either in memory, you will simply arrange for both x and y to get deleted. No problem. But then if x and y both get captured by the backwards pass (because they were used as part of some other computations), now you have no good solution, since there's no way to tell backward "hey, don't capture both of x and y".

I thought about something like `torch.utils.checkpoint.checkpoint`, so that `y` would not be kept.
After your explanation, of view vs lazy, I think something like that is probably difficult/impossible to implement.

> If it's public, linking it from this discussion would be very helpful :)

Sure, I can share some pointers to our code, but it is numpy code and the use case is audio/speech:
 - Dereverberation: https://github.com/fgnt/nara_wpe/blob/master/nara_wpe/wpe.py#L522
 - Beamforming: https://github.com/fgnt/pb_bss/blob/master/pb_bss/extraction/beamformer.py
 - Spatial distribution: https://github.com/fgnt/pb_bss/blob/master/pb_bss/distribution/complex_angular_central_gaussian.py

> In this third proposal, we also have to decide whether or not you have view semantics (mutating x.conj() affects x) or clone semantics (mutating x.conj() has no effect on x). Due to the materialized x.conj(), view semantics would be difficult, although still possible (using version counters). But it is still very tricky. For example, in your example, if z participated in backwards, then you would have ended up with both x and x.conj() saved for backwards (indeed, only conjugate views save you, at the cost of a bit more recompute).

I must say, I rarely think about inplace operations in NN code. Using no inplace operations worked for me until now.
For the `x.conj()` I hopped, that it is possible to access the actual value at the low level code (similar to conjugate view tensors), but it is kept for potential second access as temporary object attached to the python object.
Maybe this is to complicated to realize. (You called this cache for conjugate view tensors)

> Alas, I need this to work in C++ only XD. This is one place where you probably don't care, but for the good of the framework we need to make sure this is handled correctly.

Yes, I know. But I mainly write python code and I don't know much of the internals of PyTorch. So I can only guess how difficult it is to realize it in C++ and getting it working in a large framework (I try to make careful statements and I know not every python code can easily be translated, especially when threads and garbage collection is involved.).

> Yes. This is something that I am worried about too. Right now, I think it is worth it, but mostly because I think the mathematicians in the audience will get really irritated at having to translate all of their code into using explicit conjugate flags. Chances are they'll just make some wrapper object to handle it... and if they're going to do that, we might as well do it properly for them in framework.

At the moment, I must say, when I translate code to pytorch I have to do mayor changes.
For the gradient memory optimizations it would be ok to add conjugate flags.
There is already (or still) a large API difference, some operations aren't fully optimized (e.g. einsum handles manipulated strides better than matmul) and having a gradient requires also some changes.

So explicit conjugate flags, would at least make it obvious, that the optimal implementation is used.

> I do want to reiterate, that I feel that lazy conjugate views (which are implemented in this PR) are a lot simpler than lazy conjugate tensors. Their primary downside is duplicated work materializing conjugate, but as you say, if in the long term we add fusion to most operators, you will no longer have duplicated work. (Also, I suspect one could always add a cache on top later as a performance optimization; though we do have to make a call on clone or view semantics sooner rather than later.) What do you think?

At the moment, I think using `conjugate view tensors` looks like a reasonable trade off between realizable and good performance.
Compared to `conjugate lazy tensors` it doesn't change its internal state (problem with parallel and "predictable" code) and it easily fit to the gradient calculations.

> Also, I suspect one could always add a cache on top later as a performance optimization

When you add the cache, you arrive at my suggestion :), but with a better C++ implementation.

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+  //  1. Purely functional situation.  Easy: materialize all inputs and
+  //     call it a day.
+  //  2. Inplace operation.  Desugar x.add_(2) into x.conj_().add_(2).conj_().
+  //     Materialize other inputs as in (1).
```
TODO: fix this comment

anjali411 commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+  // Situations to handle:
+  //  1. Purely functional situation.  Easy: materialize all inputs and
+  //     call it a day.
+  //  2. Inplace operation.  Desugar x.add_(2) into x.conj_().add_(2).conj_().
```
why is it important to retain the conjugation bit here?

anjali411 commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+  // Situations to handle:
+  //  1. Purely functional situation.  Easy: materialize all inputs and
+  //     call it a day.
+  //  2. Inplace operation.  Desugar x.add_(2) into x.conj_().add_(2).conj_().
```
why do we have the extra `conj_()` at the end? shouldn't it just be `Desugar x.add_(2) into x.conj_().add_(2)`?

anjali411 replied:
flipping the bit is required to ensure other tensors dependent on it resolve to the correct value

anjali411 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -421,6 +421,13 @@ RegisterOperators reg(
            push(stack, arg.numel());
          },
          aliasAnalysisFromSchema()),
+     OperatorGenerator(
```
JIT support can be added in a follow-up PR!

anjali411 commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -218,6 +218,17 @@ static PyObject * THPVariable_numel(PyObject* self, PyObject* args)
    END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_is_conj(PyObject* self, PyObject* args)
```
this and some other changes can probably be skipped now by setting `manual_cpp_binding=True` in `native_functions.yaml`

anjali411 commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -1168,6 +1179,7 @@ PyMethodDef variable_methods[] = {
   {"bool", (PyCFunction)(void(*)(void))THPVariable_bool, METH_VARARGS | METH_KEYWORDS, NULL},
   {"half", (PyCFunction)(void(*)(void))THPVariable_half, METH_VARARGS | METH_KEYWORDS, NULL},
   {"int", (PyCFunction)(void(*)(void))THPVariable_int, METH_VARARGS | METH_KEYWORDS, NULL},
+  {"is_conj", (PyCFunction)THPVariable_is_conj, METH_NOARGS, NULL},
```
skip

anjali411 commented on tools/pyi/gen_pyi.py:
```diff
@@ -438,6 +438,7 @@ def gen_pyi(declarations_path, out):
         'get_default_dtype': ['def get_default_dtype() -> _dtype: ...'],
         'from_numpy': ['def from_numpy(ndarray) -> Tensor: ...'],
         'numel': ['def numel(self: Tensor) -> _int: ...'],
+        'is_conj': ['def is_conj(self: Tensor) -> _bool: ...'],
```
skip

anjali411 commented on torch/testing/__init__.py:
```diff
@@ -96,6 +96,9 @@ def _compare_tensors_internal(a: torch.Tensor, b: torch.Tensor, *, rtol, atol, e
     # (see NOTE Test Framework Tensor "Equality")
     if a.is_complex():
+        a = a.resolve_conj()
+        b = b.resolve_conj()
```
just a thought: when we modify eq and ne -->
we could check conj bit for both tensors before resolving them. We only need to resolve the tensors if only one of the two has their conj bit set.

ezyang replied:
Yes, you could write a custom Conj kernel for these, I think that's very reasonable.

anjali411 commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+
+  c10::optional<bool> is_write;
+  for (int64_t i = 0; i < num_arguments; ++i) {
+    const auto& alias_info = arguments[i].alias_info();
```
alias_info contains info about whether the tensor will get mutated in the operation

anjali411 commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+        TORCH_CHECK(*is_write == alias_info->isWrite(),
+          "Unsupported operator for conjugate fallback: ", op.schema().name(),
+          "Conjugate fallback doesn't work for operators with a mix "
+          "mutable and non-mutable inputs that alias with outputs, "
```
verify this

anjali411 committed:
Update on "Conjugate view tensor"


Based off of bdhirsch's prototype at https://github.com/pytorch/pytorch/pull/43702

Conjugate view tensor changes the semantics of x.conj() to return a view
of the original tensor, rather than immediately performing the
conjugation (aka, it implements a form of lazy evaluation for
conjugation).  This permits conjugations to be fused with later
operations that occur; e.g., matrix multiply, where BLAS libraries
typically all for a conjugate-transpose first argument.

The implementation strategy of this PR:

- Maintain conjugation status as a dispatch key in the dispatch
  key set (TODO: audit all propagation of key_set to ensure that
  we don't propagate conjugation status when we shouldn't)

- Implement a few new functions for some other operations:

  - x.is_conj() tells you whether or not the conjugate bit is
    or not.  If not x.is_conj(), what's in memory reflects
    the semantic content of the tensor; if x.is_conj(),
    the imaginary values are negated.  This is implemented
    as a fastpath function similar to numel(); see
    https://github.com/pytorch/pytorch/issues/30774

  - x.view_as_real_physical() lets you view the underlying memory
    of a possibly conjugated tensor as its actual stored
    representation.  To interpret this result correctly, you
    will typically also have to query x.is_conj()

  - x.conj_physical_() will inplace flip the sign of the
    imaginary piece of the tensor.  This is a fairly useful
    operation, especially when getting conjugations to line
    up correctly. TODO: Consider renaming this to just conj_()

  - x.resolve_conj() plays of contiguous() but for conjugate views:
    if x.is_conj(), it will return a new tensor with the conjugation
    done in memory.

- Disable view_as_real() for conjugated tensors.  This change is
  necessary because conjugation is a concept only for complex
  tensors; on real tensors, there is no way to say "view this
  memory but negate every Nth element".  Most notably, this
  means x.imag no longer works on conjugated tensors; instead
  you should do x.resolve_conj().imag instead.

- Implement a backend fallback on the conjugate dispatch key.
  This fallback works for all non-view operations.  The general
  recipe is to take each tensor argument that is_conj() and
  resolve_conj() them.  If an argument is a mutable inplace argument,
  we instead conj_physical_() inplace, set_conj(False) and then
  do our operation; on the way out, we then invert this operation.

TODO: User-level docs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

anjali411 commented on c10/core/TensorImpl.cpp:
```diff
@@ -95,6 +95,10 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
   }
 
+  if (is_conj()) {
+    TORCH_INTERNAL_ASSERT(isComplexType(typeMetaToScalarType(dtype())));
```
@ezyang why do we need this?

ezyang replied:
Internal invariant. Conjugate bit is not valid on non-complex tensors.
